imx219_modes.h

```diff
@@ -784,6 +784,95 @@ struct sensor_regs imx219_stop[] = {
 	{ 0x0100, 0x00 }, // disable streaming
 };
 
+static int imx219_set_crop(const struct sensor_def *const sensor, struct mode_def *const sensor_mode,
+			   const struct raspiraw_crop *const cfg)
+{
+	// All members of cfg default to -1 if the user did not set them.
+	// Check for validity.
+	if ((cfg->height == 0) || (cfg->width == 0))
+	{
+		// 0 is not a valid value for either.
+		return -1;
+	}
+	if ((cfg->height > 2464) || (cfg->width > 3820))
+	{
+		// These are the maximum values, can't exceed them.
+		return -1;
+	}
+	if ((cfg->left >= 0) && (cfg->left + cfg->width > 3280))
+	{
+		// Falling off the right edge of the sensor.
+		return -1;
+	}
+	if ((cfg->top >= 0) && (cfg->top + cfg->height > 3280))
+	{
+		// Falling off the bottom edge of the sensor.
+		return -1;
+	}
+
+	if (cfg->hinc >= 0)
+	{
+		modReg(sensor_mode, 0x0170, 0, 2, cfg->hinc, EQUAL);
+	}
+
+	if (cfg->vinc >= 0)
+	{
+		modReg(sensor_mode, 0x0171, 0, 2, cfg->vinc, EQUAL);
+	}
+
+	if (cfg->width > 0)
+	{
+		sensor_mode->width = cfg->width;
+		modReg(sensor_mode, 0x016C, 0, 3, cfg->width >> 8, EQUAL);
+		modReg(sensor_mode, 0x016D, 0, 7, cfg->width & 0xFF, EQUAL);
+	}
+
+	if (cfg->height > 0)
+	{
+		sensor_mode->height = cfg->height;
+		modReg(sensor_mode, 0x016E, 0, 3, cfg->height >> 8, EQUAL);
+		modReg(sensor_mode, 0x016F, 0, 7, cfg->height & 0xFF, EQUAL);
+	}
+
+	// Set the start and end pixels in both dimensions.  If the user
+	// did not provide left or top offset, then assume that we should
+	// center the image.
```

Comment:
In which case that behaviour ought to be documented in the help text, and ov5647 needs to be amended to do the same.

Reply:
I can do the help text, but I don't have a ov5647 to test these changes so I won't touch that code.

Reply:
Done.